pkg/cache/context.go

```diff
@@ -296,14 +296,21 @@ func (ctx *Context) UpdatePod(_, newObj interface{}) {
 		log.Log(log.ShimContext).Error("failed to update pod", zap.Error(err))
 		return
 	}
-	if utils.GetApplicationIDFromPod(pod) == "" {
+	applicationID := utils.GetApplicationIDFromPod(pod)
+	if applicationID == "" {
 		ctx.updateForeignPod(pod)
 	} else {
-		ctx.updateYuniKornPod(pod)
+		ctx.updateYuniKornPod(applicationID, pod)
 	}
 }
 
-func (ctx *Context) updateYuniKornPod(pod *v1.Pod) {
+func (ctx *Context) updateYuniKornPod(appID string, pod *v1.Pod) {
+	if app := ctx.getApplication(appID); app != nil {
+		if task, err := app.GetTask(string(pod.UID)); task != nil && err == nil {
+			task.setTaskPod(pod)
+		}
+	}
+
 	// treat terminated pods like a remove
 	if utils.IsPodTerminated(pod) {
 		if taskMeta, ok := getTaskMetadata(pod); ok {
```

Comment:
We already have the app, or know it exists we need to clean up here now.
The `TaskID` in the metadata is the pod UID. The appID is passed into the function already.

Reply:
I also think we need to change the signature for `notifyTaskComplete(string, string)` to `notifyTaskComplete(*Application, string)` Probably better to use a separate jira for that as it flows through into `NotifyTaskComplete()` and some tests. The 2 tests have the application pointer already. It removes a number of extra getApplication() calls we really do not need.
Similar for `ensureAppAndTaskCreated()` which is only ever called from this function. Add a parameter to it to make it: `ensureAppAndTaskCreated(*v1.Pod, *Application)` and only execute application creation `if app == nil`. This can be either in this jira or in a separate one.

Reply:
Created https://issues.apache.org/jira/browse/YUNIKORN-2724